File: test/date_tools_test.dart
import 'package:test/test.dart';
import 'package:timeasy/tools/date_tools.dart';

void main() {
  test('Week number is calculated correctly.', () {
    final dateTools = new DateTools();

    var weekNumber = dateTools.getWeekNumber(new DateTime(2021, 9, 14));
    expect(weekNumber, equals(37));

    weekNumber = dateTools.getWeekNumber(new DateTime(2020, 12, 28));
    expect(weekNumber, equals(53));

    weekNumber = dateTools.getWeekNumber(new DateTime(2021, 1, 4));
    expect(weekNumber, equals(1));
  });

  test('First day of first week id calculated correctly.', () {
    final dateTools = new DateTools();

    var firstDayOfFirstWeek = dateTools.getFirstDayOfFirstWeek(2020);
    expect(firstDayOfFirstWeek.year, 2019);
    expect(firstDayOfFirstWeek.month, 12);
    expect(firstDayOfFirstWeek.day, 30);

    firstDayOfFirstWeek = dateTools.getFirstDayOfFirstWeek(2021);
    expect(firstDayOfFirstWeek.year, 2021);
    expect(firstDayOfFirstWeek.month, 01);
    expect(firstDayOfFirstWeek.day, 04);

    firstDayOfFirstWeek = dateTools.getFirstDayOfFirstWeek(2005);
    expect(firstDayOfFirstWeek.year, 2005);
    expect(firstDayOfFirstWeek.month, 01);
    expect(firstDayOfFirstWeek.day, 03);
  });

  test('First day of week is calculated correctly', () {
    final dateTools = new DateTools();

    var firstDayOfWeek = dateTools.getFirstDayOfWeek(37, 2021);
    expect(firstDayOfWeek.year, 2021);
    expect(firstDayOfWeek.month, 9);
    expect(firstDayOfWeek.day, 13);

    firstDayOfWeek = dateTools.getFirstDayOfWeek(53, 2020);
    expect(firstDayOfWeek.year, 2020);
    expect(firstDayOfWeek.month, 12);
    expect(firstDayOfWeek.day, 28);

    firstDayOfWeek = dateTools.getFirstDayOfWeek(24, 2005);
    expect(firstDayOfWeek.year, 2005);
    expect(firstDayOfWeek.month, 6);
    expect(firstDayOfWeek.day, 13);

    firstDayOfWeek = dateTools.getFirstDayOfWeek(52, 2007);
    expect(firstDayOfWeek.year, 2007);
    expect(firstDayOfWeek.month, 12);
    expect(firstDayOfWeek.day, 24);
  });

  test('Last day of week is calculated correctly', () {
    final dateTools = new DateTools();

    var lastDayOfWeek = dateTools.getLastDayOfWeek(37, 2021);
    expect(lastDayOfWeek.year, 2021);
    expect(lastDayOfWeek.month, 9);
    expect(lastDayOfWeek.day, 19);

    lastDayOfWeek = dateTools.getLastDayOfWeek(53, 2020);
    expect(lastDayOfWeek.year, 2021);
    expect(lastDayOfWeek.month, 1);
    expect(lastDayOfWeek.day, 3);

    lastDayOfWeek = dateTools.getLastDayOfWeek(24, 2005);
    expect(lastDayOfWeek.year, 2005);
    expect(lastDayOfWeek.month, 6);
    expect(lastDayOfWeek.day, 19);

    lastDayOfWeek = dateTools.getLastDayOfWeek(52, 2007);
    expect(lastDayOfWeek.year, 2007);
    expect(lastDayOfWeek.month, 12);
    expect(lastDayOfWeek.day, 30);
  });
}

File: server/src/pkg/tools/date_tools_test.go
package tools

import (
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
)

func TestWeekNumber(t *testing.T) {
	assertWeekNumber(t, 2021, 9, 14, 37)
	assertWeekNumber(t, 2020, 12, 28, 53)
	assertWeekNumber(t, 2021, 1, 4, 1)
}

func assertWeekNumber(t *testing.T, year int, month time.Month, day int, expectedWeekNumber int) {
	weekNumber := GetWeekNumber(time.Date(year, month, day, 0, 0, 0, 0, time.UTC))
	assert.Equal(t, expectedWeekNumber, weekNumber)
}

func TestFirstDayOfFirstWeek(t *testing.T) {
	assertFirstDayOfFirstWeek(t, 2020, 2019, 12, 30)
	assertFirstDayOfFirstWeek(t, 2021, 2021, 1, 4)
	assertFirstDayOfFirstWeek(t, 2005, 2005, 1, 3)
	assertFirstDayOfFirstWeek(t, 2022, 2022, 1, 3)
	assertFirstDayOfFirstWeek(t, 2023, 2023, 1, 2)
	assertFirstDayOfFirstWeek(t, 2024, 2024, 1, 1)
	assertFirstDayOfFirstWeek(t, 2025, 2024, 12, 30)
}

func assertFirstDayOfFirstWeek(t *testing.T, year int, expectedYear int, expectedMonth time.Month, expectedDay int) {
	firstDayOfFirstWeek := GetFirstDayOfFirstWeek(year)
	assert.Equal(t, expectedYear, firstDayOfFirstWeek.Year())
	assert.Equal(t, expectedMonth, firstDayOfFirstWeek.Month())
	assert.Equal(t, expectedDay, firstDayOfFirstWeek.Day())
}

func TestFirstDayOfWeek(t *testing.T) {
	assertFirstDayOfWeek(t, 37, 2021, 2021, 9, 13)
	assertFirstDayOfWeek(t, 53, 2020, 2020, 12, 28)
	assertFirstDayOfWeek(t, 24, 2005, 2005, 6, 13)
	assertFirstDayOfWeek(t, 52, 2007, 2007, 12, 24)
	assertFirstDayOfWeek(t, 43, 2022, 2022, 10, 24)
	assertFirstDayOfWeek(t, 1, 2025, 2024, 12, 30)
}

func assertFirstDayOfWeek(t *testing.T, weekNumber int, year int, expectedYear int, expectedMonth time.Month, expectedDay int) {
	firstDayOfWeek := GetFirstDayOfWeek(weekNumber, year)
	assert.Equal(t, expectedYear, firstDayOfWeek.Year())
	assert.Equal(t, expectedMonth, firstDayOfWeek.Month())
	assert.Equal(t, expectedDay, firstDayOfWeek.Day())
}

func TestLastDayOfWeek(t *testing.T) {
	assertLastDayOfWeek(t, 37, 2021, 2021, 9, 19)
	assertLastDayOfWeek(t, 53, 2020, 2021, 1, 3)
	assertLastDayOfWeek(t, 24, 2005, 2005, 6, 19)
	assertLastDayOfWeek(t, 52, 2007, 2007, 12, 30)
}

func assertLastDayOfWeek(t *testing.T, weekNumber int, year int, expectedYear int, expectedMonth time.Month, expectedDay int) {
	lastDayOfWeek := GetLastDayOfWeek(weekNumber, year)
	assert.Equal(t, expectedYear, lastDayOfWeek.Year())
	assert.Equal(t, expectedMonth, lastDayOfWeek.Month())
	assert.Equal(t, expectedDay, lastDayOfWeek.Day())
}
